Implement Goldfish RTC

ChinYikMing opened the pull request:
# Description
This pull request introduces full RTC support, integrating a new Goldfish RTC device, userspace utilities, Buildroot updates, and comprehensive CI test coverage to ensure correctness and long-term reliability.

The Goldfish RTC device has been implemented to provide accurate system time and alarm functionality in the guest OS. At boot, the system clock is synchronized with the host’s wall clock, ensuring the guest starts with an accurate and consistent timestamp. This enables standard tools such as `date` and `hwclock` to function correctly and guarantees precise timestamps for filesystem operations like file creation and modification. Alarm events are now supported, allowing processes to schedule wake-up events via RTC interrupts. Alongside this feature, improvements were made to the DTB parsing logic to handle variable-length subnode names and to fix issues in MMIO handling when no virtio block devices are present. The Linux kernel configuration was also updated to enable the Goldfish RTC driver by default.

Buildroot was extended to include the `hwclock` utility and to support patching custom packages directly into the root filesystem. These packages are stored under `tests/system/br_pkgs/`, simplifying the workflow for testing and debugging RTC behavior while allowing additional test tools to be integrated easily in the future.

Two new userspace utilities were introduced for validating and demonstrating RTC functionality. The `rtc_alarm` program walks through enabling, triggering, and disabling RTC alarms, showing the alarm_IRQ state transition from no -> yes -> no. The `rtc_settime` utility provides a simple interface to test the `RTC_SET_TIME` ioctl by setting the RTC to either the current UTC time or a custom year supplied via command-line arguments. These tools are automatically built and included in the root filesystem to streamline testing.

Comprehensive CI test coverage was added to automatically verify RTC functionality. The workflow validates that `/dev/rtc0` is present and properly registered in `dmesg`, ensures the RTC retains updates by setting it to years both before and after the current year, and confirms that alarm interrupts behave correctly. These checks are performed using the `rtc_settime` and `rtc_alarm` utilities, verifying that time changes persist and alarm events trigger as expected. The CI process also checks that the guest’s UTC year remains aligned with the host’s wall clock. Together, these changes deliver a fully functional RTC feature set with automated testing to prevent regressions and maintain stability over time.

# Testing
1. Clone the branch
```shell
$ git clone https://github.com/ChinYikMing/rv32emu.git -b rtc --depth 1 && cd rv32emu
```

2. Build and run
```shell
$ make system ENABLE_SYSTEM=1 INITRD_SIZE=32 -j8 
```

3. Run `date`, `rtc_alarm`, `rtc_settime` in guestOS and the sample output
```shell
# date
xxx xxx xx xx:xx:xx UTC 2025

# rtc_alarm
Opening /dev/rtc0...

=== Initial /proc/driver/rtc ===
rtc_time	: 14:19:51
rtc_date	: 2025-09-21
alrm_time	: 00:00:00
alrm_date	: 1970-01-01
alarm_IRQ	: no
alrm_pending	: no
update IRQ enabled	: no
periodic IRQ enabled	: no
periodic IRQ frequency	: 1
max user IRQ frequency	: 64
24hr		: yes
===========================

Current RTC time: 2025-09-21 14:19:51 (UTC)
Setting alarm for 5 seconds later...

=== After enabling alarm ===
rtc_time	: 14:19:51
rtc_date	: 2025-09-21
alrm_time	: 14:19:56
alrm_date	: 2025-09-21
alarm_IRQ	: yes
alrm_pending	: no
update IRQ enabled	: no
periodic IRQ enabled	: no
periodic IRQ frequency	: 1
max user IRQ frequency	: 64
24hr		: yes
===========================

Alarm enabled. Waiting for it to fire...
>>> Alarm Fired! <<<

=== After alarm fired ===
rtc_time	: 14:19:56
rtc_date	: 2025-09-21
alrm_time	: 14:19:56
alrm_date	: 2025-09-21
alarm_IRQ	: no
alrm_pending	: no
update IRQ enabled	: no
periodic IRQ enabled	: no
periodic IRQ frequency	: 1
max user IRQ frequency	: 64
24hr		: yes
===========================


# rtc_settime 1980
Opening /dev/rtc0...
Setting RTC time to: 1980-01-01 00:00:00 (UTC)
RTC time successfully updated!

Reading /proc/driver/rtc to verify...

rtc_time	: 00:00:17
rtc_date	: 1980-01-01
alrm_time	: 14:19:56
alrm_date	: 2025-09-21
alarm_IRQ	: no
alrm_pending	: no
update IRQ enabled	: no
periodic IRQ enabled	: no
periodic IRQ frequency	: 1
max user IRQ frequency	: 64
24hr		: yes

```

# Expected Result
Date correctly synchronizes with the host wall clock in UTC format, alarms trigger as expected, and RTC_SET_TIME events are handled properly."
    

<!-- This is an auto-generated description by cubic. -->
---
## Summary by cubic
Add a Goldfish RTC device with host-synced UTC time and alarm interrupts, fully integrated with Linux, Buildroot, and CI. The guest now exposes /dev/rtc0, so date/hwclock work and alarms trigger via PLIC.

- **New Features**
  - Implemented Goldfish RTC MMIO (time, alarm, IRQ) with PLIC wiring; guest time syncs from rtc0 at boot.
  - Added rtc@4100000 DT node; enabled Linux RTC class and goldfish driver (HCTOSYS on rtc0).
  - Buildroot: enabled util-linux hwclock and added rtc_alarm and rtc_settime test tools via a simple package patcher.
  - CI: verifies rtc0 presence, host/guest UTC year match, RTC_SET_TIME across years, and alarm transitions (no → yes → no).
  - RTC is compile-time configurable; DTS rtc node is removed when disabled.

- **Bug Fixes**
  - DTB parsing now handles variable-length subnode names by locating '@'.
  - Virtio-blk MMIO base is computed dynamically; MMIO handling is safe when no vblk is present.
  - Corrected MASK macro to use unsigned long long for proper bit masking.

<sup>Written for commit a3a8037b2e1217f6cc7f5fa54d2f69964fc4c07d. Summary will update on new commits.</sup>

<!-- End of auto-generated description by cubic. -->

ChinYikMing committed:
Enable Goldfish RTC driver in linux config

ChinYikMing committed:
Enable hwclock utility in buildroot config

ChinYikMing committed:
Enable custom package patching for Buildroot

The do_buildroot process now supports patching from scratch, applying
the desired packages while building rootfs.cpio with the make
build-linux-image target. The desired packages should be stored under
the tests/system/br_pkgs/ .

Limitation:
The custom package is a single C source file.

TODO:
Submodule might be supported in the future. For example, Doom and Quake.

ChinYikMing committed:
Add userspace programs for /dev/rtc?

rtc_alarm.c:
The newly modified Linux configuration enables the procfs interface for
the /dev/rtc? real-time clock device. This example demonstrates how to
interact with the RTC device step by step. The program:
1. Opens the /dev/rtc? device.
2. Reads and displays its current status via the procfs interface.
3. Sets an alarm to trigger 5 seconds later.
4. Enables the alarm interrupt.
5. Reads the status again.
6. Waits for the alarm to trigger.
7. Finally reads the status once more and disables the alarm interrupt.
During this process, will observe the alarm_IRQ field transition from
no -> yes -> no, indicating that the RTC interrupt has been successfully
asserted and then handled by the emulated real time clock.

rtc_settime.c:
A simple user-space tool to test the RTC_SET_TIME ioctl on /dev/rtc?. It
sets the RTC to the current UTC system time by default or to a custom
year provided via command-line arguments. This helps verify that the
real time clock driver correctly handles time updates.

The buildroot.config is also modified to enable them by default. The
example packages are stored in the tests/system/br_pkgs/ .

ChinYikMing committed:
CI: Add RTC test coverage

This change adds CI tests to verify RTC functionality, covering the
following cases:
1. Verify that /dev/rtc0 is present and logged in dmesg.
2. Ensure /dev/rtc0 is interactable:
   - Use rtc_settime to set the year to both before(1980) and
     after(2030) the current year, then verify that the RTC retains the
     changes.
   - Use rtc_alarm to confirm that the alarm_IRQ status transitions
     correctly: no -> yes -> no.
3. Validate alignment of the UTC year between host and guest using the
   date command.

jserv reviewed (commented):
<!-- github-benchmark-action-comment(start): Benchmarks Summary -->
# Benchmarks

<details>

| Benchmark suite | Current: a3a8037b2e1217f6cc7f5fa54d2f69964fc4c07d | Previous: 95f556d0cf23b0a4178b2077526e3c73d8853d37 | Ratio |
|-|-|-|-|
| `Dhrystone` | `1626` Average DMIPS over 10 runs | `1905` Average DMIPS over 10 runs | `1.17` |
| `Coremark` | `964.623` Average iterations/sec over 10 runs | `1034.758` Average iterations/sec over 10 runs | `1.07` |

</details>

This comment was automatically generated by [workflow](https://github.com/sysprog21/rv32emu/actions?query=workflow%3ABenchmark) using [github-action-benchmark](https://github.com/marketplace/actions/continuous-benchmark).
<!-- github-benchmark-action-comment(end): Benchmarks Summary -->

ChinYikMing committed:
CI: Rebuild rootfs when new RTC custom packages are added

ChinYikMing commented on src/riscv.c:
```diff
@@ -365,6 +368,10 @@ static void load_dtb(char **ram_loc, vm_attr_t *attr)
 
+        /* set the VBLK MMIO valid range */
+        attr->vblk_mmio_base_hi = next_addr >> 20;
+        attr->vblk_mmio_max_hi = attr->vblk_mmio_base_hi + attr->vblk_cnt;
```
`attr` w/ type `vm_attr_t` is initialized with designator. From C99 description of the designator initializer: `If there are fewer initializers in a brace-enclosed list than there are elements or members of an aggregate, or fewer characters in a string literal used to initialize an array of known size than there are elements in the array, the remainder of the aggregate is subject to default initialization.` The default initialization for arithmetic type is initialized to (positive or unsigned) zero.

Thus, no need to check the `vblk_cnt` if it is zero since is redundant and lesser branch instruction in the binary.

ChinYikMing commented on src/riscv.c:
```diff
@@ -665,9 +672,8 @@ riscv_t *rv_create(riscv_user_t rv_attr)
-    attr->vblk_mmio_base_hi = 0x41;
-    attr->vblk_mmio_max_hi = attr->vblk_mmio_base_hi + attr->vblk_cnt;
+    /* setup rtc */
+    attr->rtc = rtc_new();
```
`attr` w/ type `vm_attr_t` is initialized with designator. From C99 description of the designator initializer: `If there are fewer initializers in a brace-enclosed list than there are elements or members of an aggregate, or fewer characters in a string literal used to initialize an array of known size than there are elements in the array, the remainder of the aggregate is subject to default initialization.` The default initialization for arithmetic type is initialized to (positive or unsigned) zero.

Thus, no need to check the `vblk_cnt` if it is zero since is redundant and lesser branch instruction in the binary.

visitorckw commented on src/emulate.c:
```diff
@@ -370,6 +370,8 @@ static bool has_loops = false;
 #if RV32_HAS(SYSTEM) && !RV32_HAS(ELF_LOADER)
 extern void emu_update_uart_interrupts(riscv_t *rv);
+extern void emu_update_rtc_interrupts(riscv_t *rv);
+extern uint64_t get_now_nsec(rtc_t *rtc);
```
Since this function is used to obtain the current precise time, I assume it is performance critical (please correct me if I'm wrong). The function itself is also quite short, so IMHO it would be a good candidate to move into a header as an inline function. Even if it is not made inline, since it is called from both emulate.c and rtc, I would still prefer to place the prototype in a suitable header rather than relying on extern.

jserv reviewed (changes requested):
Make RTC compile-time configurable.

ChinYikMing committed:
Make RTC compile-time configurable

- Add alias for rtc0 to decouple the memory layout changes.
- Introduce GOLDFISH_RTC for both CONFIG_ and ENABLE_ prefix (this is
  for the sake of compatability) as a selector to enable or disable to
  load the Goldfish RTC driver and device.
- Add clang-format on/off for MMIO_READ/WRITE since clang-format-20
  decreases some readability for IIF macro that used for RTC case.

visitorckw commented on src/riscv.c:
```diff
@@ -332,6 +332,16 @@ static void load_dtb(char **ram_loc, vm_attr_t *attr)
 
+/* Remove the rtc node if it is not enabled during compile time */
+#if !RV32_HAS(GOLDFISH_RTC)
+    node = fdt_path_offset(dtb_buf, "/soc@F0000000/rtc@4100000");
```
Hardcoding the DTB path seems fragile if the memory map changes. Is it possible to locate the node using its compatible string instead?

ChinYikMing replied:
> Hardcoding the DTB path seems fragile if the memory map changes. Is it possible to locate the node using its compatible string instead?

Use alias instead.

visitorckw commented on src/common.h:
```diff
@@ -29,7 +29,7 @@
 #define ARRAY_SIZE(arr) (sizeof(arr) / sizeof(arr[0]))
 
-#define MASK(n) (~((~0U << (n))))
+#define MASK(n) (~((~0ULL << (n))))
```
This modification affects the entire project. Isolating this change into a standalone prerequisite patch would help prevent unintended side effects on other components.

ChinYikMing replied:
Done, check [df80c11](https://github.com/sysprog21/rv32emu/pull/613/commits/df80c11f079b0e6188f840a44a21e9ecc9aeb0cd).

ChinYikMing committed:
Modify the integer-suffix in MASK() from U to ULL

This change enables the MASK() to manipulate 64-bit wide data.
Note that when n >= 64, the behavior of MASK() is undefined.

ChinYikMing committed:
Implement Goldfish RTC

The system date after boot is not synchronized with the host's wall
clock, resulting in incorrect timestamps shown by commands such as date.
When the system has the hwclock utility, it cannot function because
the /dev/rtc? device file is missing. In addition, programs might rely
on RTC alarm events for wake-up operations, which require access to
/dev/rtc? device file.

This commit implements an RTC device using the Goldfish RTC model,
chosen for its simplicity and native support in the Linux kernel.

By adding RTC support, the following capabilities are enabled:
- System boot-up synchronization: The RTC driver initializes the
  system time from the at boot, ensuring the guestOS starts with an
  accurate and synchronized timestamp.
- Wakeup events: The RTC can generate alarm interrupts to wake up the
  process.
- The date command, hwclock utility, and alarm configuration via ioctl
  (see the rtc(4) man page) are now functional.
- Provide accurate timestamps for filesystem operations such as when
  writing or modifying files on a virtio block device.

The #605 hardcoded vblk_mmio_base_hi to 0x41, which limits adding new
subnodes (e.g., RTC) under the SoC node. This value should instead be
dynamically set during load_dtb(). Another limitation is in the
next_addr probing logic, which currently uses subnode's name + 7. This
assumes a fixed-length name and fails for variable-length names (e.g.,
rtc@4100000, should be name + 4). The fix is to locate the '@' character
in the subnode name and use its position + 1 as the offset.
Additionally, when no virtio block device is specified,
virtio_mmio_base_hi will default to 0. In this case, use vblk_cnt for
condition checking during MMIO handling to avoid incorrect behavior.

jserv commented on src/common.h:
```diff
@@ -29,7 +29,7 @@
 #define ARRAY_SIZE(arr) (sizeof(arr) / sizeof(arr[0]))
 
-#define MASK(n) (~((~0U << (n))))
+#define MASK(n) (~((~0ULL << (n))))
```
Still undefined behavior when n >= 64. Should guard:
```c
#define MASK(n) ((n) >= 64 ? ~0ULL : ~(~0ULL << (n)))
```
Or simply ensure callers never pass n > 63.

jserv commented on tests/system/br_pkgs/rtc/rtc_settime.c:
```diff
@@ -0,0 +1,93 @@
+        new_time.tm_wday = 0;
+        new_time.tm_yday = 0;
+        new_time.tm_isdst = 0;
+
+        printf("Setting RTC time to: %04d-01-01 00:00:00 (UTC)\n", year);
+    } else {
+        /* No year provided, set RTC to current UTC time */
+        time_t now = time(NULL);
+        struct tm *utc_tm = gmtime(&now);
+
+        new_time.tm_sec = utc_tm->tm_sec;
```
No NULL check.

Consider:
```c
  struct tm *utc_tm = gmtime(&t);
  if (!utc_tm) {
      perror("gmtime failed");
      close(fd);
      return EXIT_FAILURE;
  }
```

jserv commented on src/devices/rtc.c:
```diff
@@ -0,0 +1,124 @@
+
+#include "rtc.h"
+
+static uint64_t now_nsec;
```
This global creates hidden coupling between rtc_read() calls. The TIME_HIGH read reuses now_nsec set by TIME_LOW read. If guest reads TIME_HIGH first (non-conforming but possible), it gets uninitialized/stale data.

Consider to store in rtc_t struct or compute atomically.

visitorckw replied:
The coupling is intentional and mandated by the Goldfish RTC specification [1]. The hardware requires TIME_LOW to be read first to latch the timestamp, which is then retrieved by the TIME_HIGH read.

This behavior is explicitly defined in the qemu implementation [2]:
```c
    /*
     * From the documentation linked at the top of the file:
     *
     *   To read the value, the kernel must perform an IO_READ(TIME_LOW), which
     *   returns an unsigned 32-bit value, before an IO_READ(TIME_HIGH), which
     *   returns a signed 32-bit value, corresponding to the higher half of the
     *   full value.
     */
    switch (offset) {
    case RTC_TIME_LOW:
        r = goldfish_rtc_get_count(s);
        s->time_high = r >> 32;
        r &= 0xffffffff;
        break;
    case RTC_TIME_HIGH:
        r = s->time_high;
        break;
```

BTW, for the U-Boot goldfish timer driver I am currently working on, I was explicitly requested during the review to add a comment [3] documenting this exact ordering constraint.

Link: https://android.googlesource.com/platform/external/qemu/+/refs/heads/emu-2.0-release/docs/GOLDFISH-VIRTUAL-HARDWARE.TXT [1]
Link: https://github.com/qemu/qemu/blob/master/hw/rtc/goldfish_rtc.c#L97 [2]
Link: https://lore.kernel.org/u-boot/aVTCZUacn68Nbg10@pie/ [3]

ChinYikMing replied:
Thanks @visitorckw for the reference. Added more comments to address the reuse of `now_nsec`, see [18c0635](https://github.com/sysprog21/rv32emu/pull/613/commits/18c063512633442dec6156014af65755eecbf557).

jserv commented:
In rv_check_interrupt(), called frequently
```c
  #if RV32_HAS(GOLDFISH_RTC)
      uint64_t now_nsec = rtc_get_now_nsec(PRIV(rv)->rtc);
      if (rtc_alarm_fire(PRIV(rv)->rtc, now_nsec)) {
```

clock_gettime() on every interrupt check adds syscall overhead. Consider checking only every N iterations or when alarm is actually enabled.

jserv commented on src/riscv.c:
```diff
@@ -357,7 +370,10 @@ static void load_dtb(char **ram_loc, vm_attr_t *attr)
+            char *at_pos = strchr(name, '@');
+            assert(at_pos);
+
+            uint32_t addr = strtoul(at_pos + 1, NULL, 16);
```
No endptr validation for strtoul. Malformed DTB could cause silent errors.

Consider:
```c
  char *endptr;
  uint32_t addr = strtoul(at_pos + 1, &endptr, 16);
  if (endptr == at_pos + 1) {
      rv_log_error("Invalid unit-address in node: %s", name);
      // handle error
  }
```

ChinYikMing replied:
Use this logic:
```C
char *endptr;
uint32_t addr = strtoul(at_pos + 1, &endptr, 16);
if (endptr == at_pos + 1) {
    attr->vblk_cnt = 0;
    rv_log_error(
        "Invalid unit-address in node: %s, skipping virtio blocks MMIO"
        name);
    goto dtb_end;
}
```

The lo and hi of virtio blocks MMIO address maybe malform if any node unit address is not specified, log the error message and skip the virtio blocks MMIO.

ChinYikMing committed:
Add comments for reading RTC value

jserv reviewed (changes requested):
Rebase latest `master` branch to resolve gdbstub regressions. Refine git commits.

ChinYikMing commented:
Merge this PR will rebuild the linux images, wait for #672.

ChinYikMing committed:
Modify README to reflect the support of Goldfish RTC

ChinYikMing committed:
CI: Enable Goldfish RTC for wasm system demo deployment

ChinYikMing committed:
CI: Enable Goldfish RTC for default wasm system build

ENABLE_GOLDFISH_RTC is not in the default config of wasm_defconfig, thus
need to specify it explicitly during make.

ChinYikMing commented:
Merge this PR will rebuild the linux images, blocked by #686.

jserv commented:
Thank @ChinYikMing for contributing!
